File: db4oj/core/src/com/db4o/cs/messages/MGetInternalIDs.java
/* Copyright (C) 2004   db4objects Inc.   http://www.db4o.com */

package com.db4o.cs.messages;

import com.db4o.*;
import com.db4o.cs.*;
import com.db4o.inside.*;

public final class MGetInternalIDs extends MsgD {
	public final boolean processAtServer(YapServerThread serverThread) {
		Buffer bytes = this.getByteLoad();
		long[] ids;
		synchronized (streamLock()) {
			try {
				ids = stream().getYapClass(bytes.readInt()).getIDs(transaction());
			} catch (Exception e) {
				ids = new long[0];
			}
		}
		int size = ids.length;
		MsgD message = Msg.ID_LIST.getWriterForLength(transaction(), YapConst.ID_LENGTH * (size + 1));
		Buffer writer = message.payLoad();
		writer.writeInt(size);
		for (int i = 0; i < size; i++) {
			writer.writeInt((int) ids[i]);
		}
		serverThread.write(message);
		return true;
	}
}

File: db4o.net/Db4objects.Db4o.CS/Db4objects.Db4o.CS/CS/Internal/Messages/MGetInternalIDs.cs
/* Copyright (C) 2004 - 2008  Versant Inc.  http://www.db4o.com */

using System;
using Db4objects.Db4o.CS.Internal.Messages;
using Db4objects.Db4o.CS.Internal.Objectexchange;
using Db4objects.Db4o.Foundation;
using Db4objects.Db4o.Internal;

namespace Db4objects.Db4o.CS.Internal.Messages
{
	public sealed class MGetInternalIDs : MsgD, IMessageWithResponse
	{
		public Msg ReplyFromServer()
		{
			ByteArrayBuffer bytes = GetByteLoad();
			int classMetadataID = bytes.ReadInt();
			int prefetchDepth = bytes.ReadInt();
			int prefetchCount = bytes.ReadInt();
			ByteArrayBuffer payload = MarshallIDsFor(classMetadataID, prefetchDepth, prefetchCount
				);
			MsgD message = Msg.IdList.GetWriterForLength(Transaction(), payload.Length());
			message.PayLoad().WriteBytes(payload._buffer);
			return message;
		}

		private ByteArrayBuffer MarshallIDsFor(int classMetadataID, int prefetchDepth, int
			 prefetchCount)
		{
			lock (StreamLock())
			{
				long[] ids = IdsFor(classMetadataID);
				return ObjectExchangeStrategyFactory.ForConfig(new ObjectExchangeConfiguration(prefetchDepth
					, prefetchCount)).Marshall((LocalTransaction)Transaction(), IntIterators.ForLongs
					(ids), ids.Length);
			}
		}

		private long[] IdsFor(int classMetadataID)
		{
			lock (StreamLock())
			{
				try
				{
					return Stream().ClassMetadataForID(classMetadataID).GetIDs(Transaction());
				}
				catch (Exception)
				{
				}
			}
			return new long[0];
		}
	}
}
